### DISCUSSION TITLE
Restore a backup with apps?

### DISCUSSION THREAD

**Discussion by Moonshine1439** [doronz88/pymobiledevice3#1253] [Q&A] [answered]
Hello,

I got a new iPhone 15 Pro after my old one was stolen and I'm trying to restore the backup I have on it.
I was able to restore settings and data, but all my apps are not there.
I know Apple doesn't provide a way to install apps, but can't I tell iOS which apps to download through the AppStore?

I used the following command:
```
pymobiledevice3 backup2 restore ~/Backup -p $password --settings --system --reboot --remove
```

It says at some point after the backup and reboot are complete that apps will be downloaded in the background, but none of them actually appear.
I checked the contents of the backup with pyiosbackup and I'm able to see all the apps I had on my previous iPhone.

Is there a way to do this, or do I have to manually download and rearrange all the apps as before?

Thank you in advance.

**Comment by doronz88** [comment]
From my understanding, you can only restore these apps data only after you re-download them from the AppStore. The only apps that are downloaded in the background are the system apps, which only update themselves

**Reply by Moonshine1439** [reply]
Thanks for your reply.
So if I understand correctly, you have to restore the phone like I did, then manually download all the apps you need and restore again to get the apps data back?

The backup includes all third-party apps and also Apple ones that were installed on the phone at the time of the backup, here is an example snippet from pyiosbackup:
```
'installed_apps': ['tv.twitch',
                    'ch.protonmail.calendar',
                    'com.adobe.Adobe-Reader',
                    'com.apple.Pages',
                    'com.tradingview.tradingviewapp',
                    'overlook.fing',
```

Is there nothing I can do to initiate the download from the AppStore?
When I do a restore, I have to sign in to my Apple Account again, which could maybe block the download initiation for these apps?

**Reply by doronz88** [reply]
I think you can just download the apps and only then start the restore process. Sorry I'm not familiar enough with the process myself. Only the device protocols which are exported via the USB

**Reply by Moonshine1439** [reply]
That's fine, no worries.
I was curious and looked at the code related to the backup2 restore command and I saw that it uses some options like `RestoreSystemFiles` and `RestorePreserveSettings` and they are sent to the device. Maybe there are more options to include in there, but I'm not sure how you found out about these in the first place? If you could walk me through the process of finding them, I could take a look and see if I can find something.

**Reply by doronz88** [reply]
It's not documented. You can reverse everything from `BackupAgent2`

**Reply by Moonshine1439** [reply]
I was looking around a bit today and decided to take a look at the code of `libimobiledevice` to see if I could find anything interesting.
It seems that they take special steps to restore all the applications listed in the backup by creating a `RestoreApplications.plist` file.
https://github.com/libimobiledevice/libimobiledevice/blob/master/tools/idevicebackup2.c#L552-L607

I didn't see any mention of this file in this project, so I guess it's not implemented, which explains why I was able to get my apps with idevicebackup2 when I tested it.

**Reply by doronz88** [reply, marked as answer]
Thanks for letting my know. Created a PR with this to occur by default now. Can you please test it?
https://github.com/doronz88/pymobiledevice3/pull/1259

**Reply by Moonshine1439** [reply]
I just finished a restore using the exact command I used in the first post and I was able to get all my apps in a "waiting to download" state to their correct location (grid placement/folder). Thank you very much!
